### ISSUE TITLE
Respond with an error on no connection to downstream service

### ISSUE THREAD

**Issue by Deilan** [ThreeMammals/Ocelot#603]
## New Feature
Respond with a reasonable error message in a case when _No connection could be made because the target machine actively refused it_. And add an extension point to override the message and status code (e. g. 502 Bad Gateway) for this case of course.

The error usually looks like the following in logs:
```
warn: Ocelot.Requester.Middleware.HttpRequesterMiddleware[0]
      requestId: 0HLGKQK19POEE:00000002, previousRequestId: no previous request id, message: Error making http request, exception: System.Net.Http.HttpRequestException: No connection could be made because the target machine actively refused it ---> System.Net.Sockets.SocketException: No connection could be made because the target machine actively refused it
         at System.Net.Http.ConnectHelper.ConnectAsync(String host, Int32 port, CancellationToken cancellationToken)
         --- End of inner exception stack trace ---
         at System.Net.Http.ConnectHelper.ConnectAsync(String host, Int32 port, CancellationToken cancellationToken)
         at System.Threading.Tasks.ValueTask`1.get_Result()
         at System.Net.Http.HttpConnectionPool.CreateConnectionAsync(HttpRequestMessage request, CancellationToken cancellationToken)
         at System.Threading.Tasks.ValueTask`1.get_Result()
         at System.Net.Http.HttpConnectionPool.WaitForCreatedConnectionAsync(ValueTask`1 creationTask)
         at System.Threading.Tasks.ValueTask`1.get_Result()
         at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
         at System.Net.Http.DiagnosticsHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
         at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
         at Ocelot.Requester.HttpClientHttpRequester.GetResponse(DownstreamContext context)
```

## Motivation for New Feature
More smooth UX.

## Steps to Reproduce the Problem
Reroute to a host/port, which doesn't run a web server.

**Comment by TomPallister** [comment]
@Deilan what would you suggest for the error message?

**Comment by Deilan** [comment]
@TomPallister 
"_No connection could be made because the target machine actively refused it_" would be pretty good,
This error does get returned to users, the message and status could be overrided, and message could be omitted at all (like how it works now).

**Comment by TomPallister** [comment]
@Deilan can you give me an example of where this error gets returned to a user? I cannot replicate the issue! When Ocelot records an error it appends to a list of errors internally and then before returning it logs the errors, maps the errors into a status code and returns an empty response.

**Comment by Deilan** [comment]
@TomPallister It's a feature request, not a bug report. Currently the user just receives an empty body and 500 status code.

**Comment by TomPallister** [comment]
@Deilan I understand that it is a feature request but I don't have much free time so I try and get straight to the point! 

I don't fully understand what it is you are suggesting.

I get that you want a smoother error message, but in the logs surely you want everything? Not just the message e.g. "No connection could be made because the target machine actively refused it", you want the stack etc. Which is what we currently have.

Are you saying when we get a 500 there should be an extension point that lets the developer specify what should be returned to users? This would make sense, friendly error messages for users of an API. 

I'm not sure about what error codes should be returned in what circumstances. If there is an error calling a downstream service I just want to return a 500, I'm not sure how you know this should be a 502? The users request is to Ocelot so if Ocelot gets an error calling downstream it is just a server error really? Ocelot worked fine it was calling Downstream that didn't work?

We should probably look at how other API gateways handle this.

**Comment by thomaslevesque** [comment]
Any news on this? This is a pretty important thing for a gateway to implement properly.
Returning **500 Internal Error** tells nothing useful to the client; Ocelot should really return **502 Bad Gateway** when it can't reach the downstream server.

**Comment by dalyl** [comment]
I have more than one machine, one of which is out of order, how to find the available machine automatically（Using static routing tables）

**Comment by TomPallister** [comment]
fixed by #1149
